packages/vite/src/node/plugins/vendorAnalysis.ts

```diff
@@ -0,0 +1,244 @@
+import type { ResolvedConfig } from '../config'
+import type { Plugin } from '../plugin'
+import type { ModuleInfo } from 'rollup'
+import type { ImportSpecifier } from 'es-module-lexer'
+import { init, parse as parseImports } from 'es-module-lexer'
+import { isCSSRequest } from './css'
+import { createDebugger, timeFrom } from '../utils'
+import { performance } from 'perf_hooks'
+
+const isDebug = !!process.env.DEBUG
+const debug = createDebugger('vite:vendor-analysis')
+
+/**
+ * Build only. Analyze which modules are used by entries and
+ * generate manualChunks config to split those files to vendor.
+ */
+export function buildVendorAnalysisPlugin(config: ResolvedConfig): Plugin {
+  const vendorImports = new Set<string>()
+  return {
+    name: 'vite:vendor-analysis',
+    async buildEnd() {
+      await init
+
+      const entrieIds = new Set<string>()
+      const importsCache = new Map<string, readonly ImportSpecifier[]>()
+      const exportsCache = new Map<string, readonly string[]>()
+      const parseImps = async (
+        info: ModuleInfo,
+        trace: string[]
+      ): Promise<[readonly ImportSpecifier[], readonly string[]]> => {
+        // circular dependency
+        if (trace.includes(info.id))
+          return [
+            importsCache.get(info.id) || [],
+            exportsCache.get(info.id) || []
+          ]
+        trace.push(info.id)
+
+        const [imports, exps] = parseImports(info.code!)
+        const exports = [...exps]
+
+        // process export *
+        for (const imp of imports) {
+          // don't process dynamic imports
+          if (imp.d !== -1 || !imp.n) continue
+          // resolve the absolute id
+          const id = (await this.resolve(imp.n, info.id))?.id
+          if (!id) continue
+          const rinfo = this.getModuleInfo(id)
+          if (!rinfo || !rinfo.code) continue
+          const statment: string = info.code!.substring(imp.ss, imp.se)
+          if (
+            statment.startsWith('export *') &&
+            !statment.startsWith('export * as')
+          ) {
+            // merge all exports from module
+            ;(await findExports(rinfo, trace)).forEach((exp) => {
```

Comment:
is this leading semicolon meant to be here?

Reply:
prettier just did that, but I just rewrote that so not an issue anymore.